### ISSUE TITLE
"No proc named" for existing proc

### ISSUE THREAD

**Issue by Hinaichigo** [OpenDreamProject/OpenDream#704]
Trying to compile /vg/station, the following error occurs:
```
Error at code\names.dm:1:31: No proc named "file2list"
```
The relevant code in `names.dm`:
```
var/list/ai_names = file2list("config/names/ai.txt")
```
file2list is defined as:
```
/proc/file2list(filename, seperator="\n")
	return splittext(return_file_text(filename),seperator)
```
Maybe it has something to do with there being a number in the proc name, because it also occurs for the proc:
`file2listExceptComments`
defined as:
```
/proc/file2listExceptComments(filename, seperator="\n", comment_prefix="#", trim_all=TRUE)
	. = list()
	var/prefixlen=length(comment_prefix)
	for(var/line in file2list(filename,seperator))
		var/trim_line=trim(line)
		// Drop empty lines.
		if(trim_line == "")
			continue
		// Drop comments.
		if(copytext(trim_line,1,1+prefixlen)==comment_prefix)
			continue
		. += (trim_all ? trim_line : line)
```

**Comment by wixoaGit** [comment]
> Maybe it has something to do with there being a number in the proc name

The compiler correctly handles identifiers that contain numbers, so that's not the issue.

The issue is that the compiler is attempting to use the `file2list()` proc before it gets defined. Currently it resolves the values of global vars as soon as it encounters them, and `code/names.dm` is compiled before `code/__HELPERS/type2type.dm` where the proc is defined.

**Comment by Hinaichigo** [comment]
So processing `_` earlier than alphanumerics would fix it? That seems like the it could be the reason why the folder is  `__HELPERS` instead of just `helpers`. DM has a similar issue with `#define`s, but that might be less avoidable.

**Comment by wixoaGit** [comment]
The compiler already processes the files in the order they're `#include`d in the DME, which is (typically) already set up by the Dream Maker software. Dream Maker includes the files alphabetically with `_` coming first, but it also includes all the files in a folder before moving on to subdirectories.

Fixing the issue would require postponing the step of resolving global var values until after every global proc has been defined.

**Comment by ike709** [comment]
Dupe of another issue

**Comment by wixoaGit** [comment]
This isn't a dupe of any issue

**Comment by Hinaichigo** [comment]
There's also an issue where:
```
40 var/wires = WIRE_RECEIVE | WIRE_PULSE
41 
42 var/const/WIRE_RECEIVE = 1			//Allows Pulsed(0) to call Activate()
43 var/const/WIRE_PULSE = 2				//Allows Pulse(0) to act on the holder
```
errors `Error at code\modules\assembly\assembly.dm:40:39: Unknown identifier "WIRE_PULSE"`
Is this caused by the same or a similar thing? It seems like maybe BYOND defines all `const`s before evaluating the value of non-`const` vars.

**Comment by Hinaichigo** [comment]
Am I on the right track in thinking that order to fix this:
```
// TODO Nuke this pass
foreach (DMObject dmObject in DMObjectTree.AllObjects) {
    dmObject.CreateInitializationProc();
}

foreach (DMProc proc in DMObjectTree.AllProcs)
    proc.Compile();
```
in `DMObjectBuilder.cs` would need to be changed to put all of the unparsed var definitions into a list for all files, and then once all files are processed, go through that list and evaluate them?

**Comment by wixoaGit** [comment]
Taking a deeper look at this, this is actually relatively easy to fix. The only thing that tries to resolve a var's proc call before it can be defined is [this line](https://github.com/wixoaGit/OpenDream/blob/1241a9db75ed8f24400023f519287e0ffc7f773f/DMCompiler/DM/Visitors/DMObjectBuilder.cs#L234), which could be replaced with a method that gets the proc's name without ever resolving the proc itself.

> There's also an issue where:
> 
> ```
> 40 var/wires = WIRE_RECEIVE | WIRE_PULSE
> 41 
> 42 var/const/WIRE_RECEIVE = 1			//Allows Pulsed(0) to call Activate()
> 43 var/const/WIRE_PULSE = 2				//Allows Pulse(0) to act on the holder
> ```
> 
> errors `Error at code\modules\assembly\assembly.dm:40:39: Unknown identifier "WIRE_PULSE"` Is this caused by the same or a similar thing? It seems like maybe BYOND defines all `const`s before evaluating the value of non-`const` vars.

It's caused by a similar issue, though more difficult to fix. The OD compiler is trying to resolve the var's value immediately instead of waiting until every const is defined.

The solution for this is to postpone setting these variables' values until after [this `ProcessBlockInner()` call](https://github.com/wixoaGit/OpenDream/blob/1241a9db75ed8f24400023f519287e0ffc7f773f/DMCompiler/DM/Visitors/DMObjectBuilder.cs#L30). What makes this complicated is knowing when the var depends on a const that doesn't exist yet, and postponing it until after it is. Because of code snippets like this, it could even end up taking multiple iterations:
```
/var/const/A = B
/var/const/B = C
/var/const/C = D
/var/const/D = 1
```

**Comment by Hinaichigo** [comment]
Trying to compile that in BYOND results in:
```
code\test.dm:1:warning: B: use of B precedes its definition
code\test.dm:2:warning: B: definition is here
code\test.dm:2:warning: C: use of C precedes its definition
code\test.dm:3:warning: C: definition is here
code\test.dm:3:warning: D: use of D precedes its definition
code\test.dm:4:warning: D: definition is here
```
So maybe it wouldn't have be iterative.

**Comment by wixoaGit** [comment]
Yes, but it _does_ work, interestingly enough.
![image](https://user-images.githubusercontent.com/30789242/174523755-da5e3694-63fd-4c7a-9f8e-6efdc9a9835b.png)

**Comment by Hinaichigo** [comment]
This no longer occurs and seems like it was fixed by #872.
